carddav/match.go

```diff
@@ -0,0 +1,149 @@
+package carddav
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/emersion/go-vcard"
+)
+
+// Filter returns the filtered list of address objects matching the provided query.
+// A nil query will return the full list of address objects.
+func Filter(query *AddressBookQuery, aos []AddressObject) ([]AddressObject, error) {
+	if query == nil {
+		// FIXME: should we always return a copy of the provided slice?
+		return aos, nil
+	}
+
+	n := query.Limit
+	if n <= 0 {
+		n = len(aos)
+	}
+	out := make([]AddressObject, 0, n)
+	for _, ao := range aos {
+		ok, err := Match(query, ao)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			continue
+		}
+		out = append(out, ao)
+		if len(out) >= n {
+			break
+		}
+	}
+	return out, nil
+}
+
+// Match reports whether the provided AddressObject matches the query.
+func Match(query *AddressBookQuery, ao AddressObject) (matched bool, err error) {
+	if query == nil {
+		return true, nil
+	}
+
+	if query.DataRequest.AllProp {
```

Comment:
`AllProp` means that the client wants all properties to be returned, ie. `Props` is ignored. Since we don't filter returned properties yet (we always return all props), we can ignore it.

This can probably be turned into a TODO comment.